### PR TITLE
#86 Show overlay to everyone if the game is being hard synced.

### PR DESCRIPTION
**Pull request by zuev93**
Also replaced most of EventHandlers since their are redundant.

### CODE REVIEW

**Comment by polycone** [comment]
Could you please make separate changes in separate commits in the future? Pretty hard to follow the mixed state of refactored events and the implementation.

**Comment by zuev93** [comment]
> Could you please make separate changes in separate commits in the future? Pretty hard to follow the mixed state of refactored events and the implementation.

Reshufled changes between PRs. Now this should be better

File: MultiplayerMod/Multiplayer/State/MultiplayerState.cs [hunk 1]
```diff
@@ -1,9 +1,14 @@
...
 using System.Collections.Generic;
+using MultiplayerMod.Core.Extensions;
 
 namespace MultiplayerMod.Multiplayer.State;
 
 [Serializable]
 public class MultiplayerState {
     public Dictionary<IPlayer, PlayerState> Players = new();
+
+    public void Rebuild() {
```
Comment (polycone) [inline MultiplayerMod/Multiplayer/State/MultiplayerState.cs:11]: A bit misleading as of now, maybe it's better to reset spawned state in the place? The function is too narrow and tightly related to hard sync logic.

File: MultiplayerMod/Multiplayer/State/MultiplayerGame.cs [hunk 2]
```diff
@@ -4,14 +4,24 @@ namespace MultiplayerMod.Multiplayer.State;
 
 public static class MultiplayerGame {
 
+    public static IPlayer Player { get; set; } = null!;
```
Comment (polycone) [inline MultiplayerMod/Multiplayer/State/MultiplayerGame.cs:7]: Hmmm, now we have two sources of truth.
It was supposed to use `IMultiplayerClient.Player` to get the current player.

File: MultiplayerMod/Multiplayer/State/MultiplayerGame.cs [hunk 3]
```diff
@@ -4,14 +4,24 @@ namespace MultiplayerMod.Multiplayer.State;
...
     public static MultiplayerState State { get; set; } = new();
-    public static MultiplayerObjects Objects { get; set; } = new();
+    public static MultiplayerObjects Objects { get; private set; } = new();
+
+    public static PlayerState CurrentPlayerState {
+        get {
+            if (State.Players.TryGetValue(Player, out var state))
+                return state;
+
+            State.Players.Add(Player, new PlayerState(Player));
```
Comment (polycone) [inline MultiplayerMod/Multiplayer/State/MultiplayerGame.cs:17]: Shouldn't be this an illegal case when the current player doesn't have state🤔

File: MultiplayerMod/Multiplayer/State/MultiplayerGame.cs [hunk 4]
```diff
@@ -4,14 +4,24 @@ namespace MultiplayerMod.Multiplayer.State;
...
+            return State.Players[Player];
```
Comment (polycone) [inline MultiplayerMod/Multiplayer/State/MultiplayerGame.cs:18]: It was supposed to use it like that in place, just `State.Players[client.Player]`.

File: MultiplayerMod/Multiplayer/State/MultiplayerGame.cs [hunk 5]
```diff
@@ -4,14 +4,24 @@ namespace MultiplayerMod.Multiplayer.State;
...
 public static class MultiplayerGame {
 
+    public static IPlayer Player { get; set; } = null!;
     public static MultiplayerRole Role { get; set; }
-    public static bool WorldSpawned { get; set; }
     public static MultiplayerState State { get; set; } = new();
-    public static MultiplayerObjects Objects { get; set; } = new();
+    public static MultiplayerObjects Objects { get; private set; } = new();
+
+    public static PlayerState CurrentPlayerState {
```
Comment (polycone) [inline MultiplayerMod/Multiplayer/State/MultiplayerGame.cs:12]: Maybe it's ok to add something like 
```csharp
    public PlayerState Current => Players[Container.Get<IMultiplayerClient>().Player];
```
into the state itself.

File: MultiplayerMod/Multiplayer/State/PlayerState.cs [hunk 6]
```diff
@@ -7,6 +7,7 @@ namespace MultiplayerMod.Multiplayer.State;
 public class PlayerState {
     public IPlayer Player { get; }
     public Vector2 CursorPosition { get; set; }
+    public bool Spawned { get; set; }
```
Comment (polycone) [inline MultiplayerMod/Multiplayer/State/PlayerState.cs:10]: From the usage perspective seems like it shows whether a player is spawned. Maybe WorldSpawned?
`playerState.Spawned` vs `playerState.WorldSpawned`

File: MultiplayerMod/Network/IMultiplayerServer.cs [hunk 7]
```diff
@@ -17,6 +17,7 @@ public interface IMultiplayerServer {
     void Send(IMultiplayerCommand command, MultiplayerCommandOptions options = MultiplayerCommandOptions.None);
 
     event Action<ServerStateChangedEventArgs> StateChanged;
-    event Action<PlayerConnectedEventArgs> PlayerConnected;
+    event Action<PlayerEventArgs> PlayerConnected;
+    event Action<PlayerEventArgs> PlayerDisconnected;
```
Comment (polycone) [inline MultiplayerMod/Network/IMultiplayerServer.cs:21]: Just `<IPlayer>`? Since we're removing unnecessary args.

**Review by polycone — APPROVED** [review]

File: MultiplayerMod/Multiplayer/Configuration/GameEventBindings.cs [hunk 8]
```diff
@@ -91,6 +91,10 @@ public class GameEventBindings {
             containers => client.Send(new InitializeImmigration(containers));
         ImmigrantScreenEvents.Proceed += deliverable => client.Send(new ProceedImmigration(deliverable));
         ImmigrantScreenEvents.Reject += () => client.Send(new RejectImmigration());
+        PauseScreenEvents.QuitGame += () => {
+            if (client.State >= MultiplayerClientState.Connecting)
```
[the remark this answers is not in the document]
Comment (zuev93) [reply MultiplayerMod/Multiplayer/Configuration/GameEventBindings.cs:95]: We could. But this is a refactoring and this line have been moved from MultiplayerCoordinator.

So I'd like to keep this out of a scope for current PR.

File: MultiplayerMod/Multiplayer/State/MultiplayerGame.cs (hunk 5 above)
Comment (zuev93) [reply MultiplayerMod/Multiplayer/State/MultiplayerGame.cs:12]: Done

File: MultiplayerMod/Network/IMultiplayerServer.cs (hunk 7 above)
Comment (zuev93) [reply MultiplayerMod/Network/IMultiplayerServer.cs:21]: Done
